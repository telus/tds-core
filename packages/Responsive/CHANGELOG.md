# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="1.1.0"></a>
# [1.1.0](https://github.com/telusdigital/tds/compare/@tds/core-responsive@1.0.0...@tds/core-responsive@1.1.0) (2018-04-02)


### Bug Fixes

* **core-display-heading:** remove unnecessary variable ([4a48ed4](https://github.com/telusdigital/tds/commit/4a48ed4))
* **shared-base-button:** refactor for sass based responsiveness ([a17f6ac](https://github.com/telusdigital/tds/commit/a17f6ac))


### Features

* **core-responsive:** move responsive scss file to root ([97924b8](https://github.com/telusdigital/tds/commit/97924b8))
