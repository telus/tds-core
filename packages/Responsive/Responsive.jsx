import React from 'react'
import PropTypes from 'prop-types'
import Media from 'react-media'

import { warn } from '../../shared/utils/warn'

const breakpoints = {
  sm: 576,
  md: 768,
  lg: 992,
  xl: 1200,
}

/**
 * Respond to device features, most commonly the browser viewport size.
 *
 * @version ./package.json
 */
const Responsive = ({ minWidth, maxWidth, query, children, ...rest }) => {
  if (!minWidth && !maxWidth) {
    warn('Responsive', 'Responsive needs a minWidth or maxWith prop')
  }

  // Do it this way to not create an object where some keys have the value "undefined", which causes the generated media-query to be invalid.
  // (min-width: 300px) and (max-width: undefined) is bad :(
  // (min-width: 300px) is good :)
  const mediaQuery = {}
  if (minWidth) {
    mediaQuery.minWidth = breakpoints[minWidth]
  }
  if (maxWidth) {
    mediaQuery.maxWidth = breakpoints[maxWidth] - 1
  }

  return (
    <Media {...rest} query={{ ...mediaQuery, ...query }}>
      {children}
    </Media>
  )
}

Responsive.propTypes = {
  /**
   * Set a min-width media query.
   */
  minWidth: PropTypes.oneOf(['sm', 'md', 'lg', 'xl']),
  /**
   * Set a max-width media query.
   */
  maxWidth: PropTypes.oneOf(['sm', 'md', 'lg', 'xl']),
  /**
   * An object containing any valid CSS media query characteristics. It will be converted to a CSS media query.
   * Use `query` if you need characteristics other than `minWidth` and `maxWidth`.
   *
   * Any `minWidth` or `maxWidth` values here will override the props by the same name.
   */
  query: PropTypes.object,
  /**
   * The content. Can be text, any HTML element, a function, or any component.
   */
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
}

Responsive.defaultProps = {
  minWidth: undefined,
  maxWidth: undefined,
  query: {},
  children: undefined,
}

export default Responsive
