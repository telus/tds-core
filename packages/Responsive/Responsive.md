The `Responsive` component is a thin wrapper over the [**react-media**](https://github.com/ReactTraining/react-media) community
component, which "listens for matches to a CSS media query and renders its child component based on whether the query matches or not."

This package also contains Sass media query methods (documented below). In most cases **we recommend using the Sass methods
over the Responsive React component** since they do not adversely affect server-side rendering. While not recommended for general use, the Responsive component is useful if you want to do something programmatically. For example, changing a prop or swapping a component on resize.

### Breakpoints

The included Sass-based media queries and the `Responsive` component both use the same breakpoint keys. The following chart details these breakpoint keys, and what values they represent.

| Breakpoint key | `min-width` / `$from` Value | `max-width` / `$until` Value |
| :------------- | :-------------------------- | :--------------------------- |
| `sm`           | 576px                       | 575px                        |
| `md`           | 768px                       | 767px                        |
| `lg`           | 992px                       | 991px                        |
| `xl`           | 1200px                      | 1199px                       |

### Responsive component usage

Use the `render` prop for the common case of rendering a component only when the media query matches.

Use the `minWidth` and `maxWidth` props to configure viewport size ranges. The breakpoint specifiers will be converted to pixel-based
media queries.

```jsx
<div>
  <Responsive minWidth="sm" maxWidth="md" render={() => <Text>The document is at "small".</Text>} />
  <Responsive minWidth="md" render={() => <Text>The document is at "medium" or above.</Text>} />
</div>
```

For more power use a child function, whose only argument will be a boolean flag that indicates whether the media query
matches or not.

```jsx
<Responsive minWidth="md">
  {matches =>
    matches ? (
      <Text>The document is at "medium" or above.</Text>
    ) : (
      <Text>The document is less than "medium".</Text>
    )
  }
</Responsive>
```

Use the `query` prop if you need to match on [other media features](https://developer.mozilla.org/en-US/docs/Web/CSS/@media#Media_features).

#### defaultMatches prop for server-side rendering

This component comes with a `defaultMatches` prop and its default is set to true.

When rendering on the server you can use the `defaultMatches` prop to set the initial state on the server to match whatever you think it will be on the client.

You can detect the user's device by analyzing the user-agent string from the HTTP request in your server-side rendering code. There are many ways of doing this, a popular tool is [mobile-detect](https://www.npmjs.com/package/mobile-detect).

```jsx
initialState = {
  device: 'desktop', // add your own detection logic here
}
;<div>
  <Responsive
    maxWidth="md"
    defaultMatches={state.device === 'mobile'}
    render={() => <Text>Render me below medium breakpoint.</Text>}
  />

  <Responsive
    minWidth="md"
    defaultMatches={state.device === 'desktop'}
    render={() => <Text>Render me above medium breakpoint.</Text>}
  />
</div>
```

#### Testing responsive behaviour

Moving the responsive behaviour into JavaScript enables testing of the results of media queries, which is impossible with CSS-based media queries. It is straightforwards to mock or stub the result of a media query for testing environments. Here are some examples of how we incorporated unit tests for responsive behaviour in some of the TDS components:

- [Button component tests](https://github.com/telusdigital/tds-core/blob/b2108d1074383ba887c5b87a2c3055799937fcd3/packages/Button/__tests__/Button.spec.jsx#L52-L68) and its corresponding [JSX](https://github.com/telusdigital/tds-core/blob/b2108d1074383ba887c5b87a2c3055799937fcd3/shared/components/BaseButton/BaseButton.jsx#L17-L34) using the matches boolean flag
- [Tooltip component tests](https://github.com/telusdigital/tds-core/blob/b2108d1074383ba887c5b87a2c3055799937fcd3/packages/Tooltip/__tests__/Tooltip.spec.jsx#L56-L102) and its corresponding [JSX](https://github.com/telusdigital/tds-core/blob/b2108d1074383ba887c5b87a2c3055799937fcd3/packages/Tooltip/Tooltip.jsx#L81-L108) rendering a particular bubble only when the media query matches

To implement responsive behaviour in your component, this package provides Sass-based media queries using [**sass-mq**](https://github.com/sass-mq/sass-mq), and a standardized set of breakpoints.

The `Responsive` React component is a thin wrapper over the [**react-media**](https://github.com/ReactTraining/react-media) community component, which "listens for matches to a CSS media query and renders its child component based on whether the query matches or not." While it is available, this component is not recommended for most use-cases.

### Sass media query method usage

To make use of Sass media queries through the Responsive component, first declare the following import at the top of your `.scss` file.

```scss
@import '~@tds/core-responsive/responsive';
```

Once this is done, you may leverage [**sass-mq**](https://github.com/sass-mq/sass-mq) style media queries to add responsive behaviour to your component.

```scss
.someClass {
  @include mq($from: sm, $until: lg) {
    width: 500px;
  }
  @include mq($from: xl) {
    width: 900px;
  }
}

.someOtherClass {
  @include mq($until: md, $and: '(orientation: landscape)') {
    width: 100%;
    height: 100%;
  }
}
```

#### MQ Parameters

| Parameter | Values                             | Effect                                                                                                                                  |
| --------- | ---------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------- |
| `$from`   | `sm`, `md`, `lg`, `xl`             | The CSS rules contained within will start at the defined point.                                                                         |
| `$until`  | `sm`, `md`, `lg`, `xl`             | The CSS rules contained within will end at the defined point.                                                                           |
| `$and`    | Standard CSS Media Query Arguments | Allows for custom arguments like orientation or DPI specifications using the standard CSS syntax. These are encased in quotation marks. |
