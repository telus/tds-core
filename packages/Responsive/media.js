export const breakpoints = {
  sm: 576,
  md: 768,
  lg: 992,
  xl: 1200,
}

export default {
  query: {},
  from(breakpoint) {
    if (breakpoint !== 'xs') {
      this.query.minWidth = breakpoint
    }
    return this
  },
  until(breakpoint) {
    this.query.maxWidth = breakpoint
    return this
  },
  and(custom) {
    this.query.and = custom
    return this
  },
  css(style) {
    const { minWidth, maxWidth, and } = this.query

    const min = minWidth ? `(min-width: ${breakpoints[minWidth]}px)` : undefined
    const max = maxWidth ? `(max-width: ${breakpoints[maxWidth] - 1}px)` : undefined

    if (typeof min !== 'undefined' || typeof max !== 'undefined' || typeof and !== 'undefined') {
      const mediaQuery = `@media ${[min, max, and].filter(a => a).join(' and ')}`
      this.query = {}

      return {
        [mediaQuery]: {
          ...(typeof style === 'function' ? style() : style),
        },
      }
    }

    return typeof style === 'function' ? style() : style
  },
}
