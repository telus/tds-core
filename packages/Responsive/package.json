{
  "name": "@tds/core-responsive",
  "version": "1.3.3",
  "description": "Respond to device features, most commonly the browser viewport size",
  "main": "index.cjs.js",
  "module": "index.es.js",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/telusdigital/tds.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "TELUS digital",
  "engines": {
    "node": ">=8"
  },
  "bugs": {
    "url": "https://github.com/telusdigital/tds/issues"
  },
  "homepage": "http://tds.telus.com",
  "peerDependencies": {
    "react": ">=15",
    "react-dom": ">=15"
  },
  "dependencies": {
    "prop-types": "^15.5.10",
    "sass-mq": "^5.0.0",
    "react-media": "^1.10.0"
  }
}
