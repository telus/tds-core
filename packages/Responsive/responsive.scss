$mq-breakpoints: (
  sm: 576px,
  md: 768px,
  lg: 992px,
  xl: 1200px
);

@import '~sass-mq/mq';
