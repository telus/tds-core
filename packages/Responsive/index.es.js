import Responsive, { media } from './dist/index.es'

export { media }
export default Responsive
