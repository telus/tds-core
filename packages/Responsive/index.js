import Responsive from './Responsive'
import media from './media'

export { media }
export default Responsive
