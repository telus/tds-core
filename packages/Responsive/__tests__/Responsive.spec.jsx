import React from 'react'
import { shallow } from 'enzyme'

import Responsive from '../Responsive'
import { warn } from '../../../shared/utils/warn'

jest.mock('../../../shared/utils/warn')

describe('Responsive with react-media', () => {
  const defaultProps = {
    minWidth: 'sm',
  }
  const defaultChildren = 'Some content'

  const doShallow = (props = defaultProps, children = defaultChildren) =>
    shallow(<Responsive {...props}>{children}</Responsive>)

  it('renders', () => {
    const responsive = doShallow()

    expect(responsive).toMatchSnapshot()
  })

  // TODO: Maybe this warning should not exist? What is you just want to query for screen vs print or something?
  it('warns when min width and max width are not defined', () => {
    doShallow({ minWidth: undefined, maxWidth: undefined })

    expect(warn).toHaveBeenCalled()
  })

  it('translates min and max width to pixel breakpoints', () => {
    const responsive = doShallow({ minWidth: 'sm', maxWidth: 'md' })

    expect(responsive).toHaveProp('query', { minWidth: 576, maxWidth: 767 })
  })

  it('does not generate a query with undefined min or max widths', () => {
    let responsive = doShallow({ minWidth: 'sm' })

    let query = responsive.props().query
    expect(query.hasOwnProperty('maxWidth')).toBeFalsy() // eslint-disable-line no-prototype-builtins

    responsive = doShallow({ maxWidth: 'sm' })

    query = responsive.props().query
    expect(query.hasOwnProperty('minWidth')).toBeFalsy() // eslint-disable-line no-prototype-builtins
  })

  it('accepts other media query characteristics', () => {
    const responsive = doShallow({ query: { screen: true, orientation: 'portrait' } })

    expect(responsive).toHaveProp(
      'query',
      expect.objectContaining({ screen: true, orientation: 'portrait' })
    )
  })

  it('allows overriding the min and max width with a query', () => {
    const responsive = doShallow({ query: { minWidth: 100, maxWidth: 2000 } })

    expect(responsive).toHaveProp('query', { minWidth: 100, maxWidth: 2000 })
  })

  it('passes additional attributes to the Media component', () => {
    const responsive = doShallow({ defaultMatches: false })

    expect(responsive).toHaveProp('defaultMatches', false)
  })
})
