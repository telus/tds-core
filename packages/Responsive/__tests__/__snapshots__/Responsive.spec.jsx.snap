// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Responsive with react-media renders 1`] = `
<Media
  defaultMatches={true}
  query={
    Object {
      "minWidth": 576,
    }
  }
>
  Some content
</Media>
`;
